src/WebJobs.Script/Diagnostics/FileTraceWriter.cs

```diff
@@ -39,90 +40,141 @@ public FileTraceWriter(string logFilePath, TraceLevel level, LogType logType) :
+                if (_flushTimer == null)
+                {
+                    // start a timer to flush accumulated logs in batches
+                    _flushTimer = new Timer(FlushCallback, null, LogFlushIntervalMs, LogFlushIntervalMs);
+                }
             }
+        }
 
-            if (currentBuffer.Count == 0)
+        private protected virtual void StopFlushTimer()
+        {
+            if (_flushTimer == null)
             {
                 return;
             }
 
-            // concatenate all lines into one string
-            StringBuilder sb = new StringBuilder();
-            string line = null;
-            while (currentBuffer.TryDequeue(out line))
+            lock (_flushTimerSyncRoot)
             {
-                sb.AppendLine(line);
+                if (_flushTimer != null)
+                {
+                    _flushTimer.Dispose();
+                    _flushTimer = null;
+                }
             }
+        }
 
-            if (_currentLogFileInfo == null)
-            {
-                // delay create log file
-                SetLogFile();
-            }
+        private void FlushCallback(object state)
+        {
+            Flush(true);
+        }
 
-            // write all lines in a single file operation
-            string content = sb.ToString();
-            try
-            {
-                AppendToFile(_currentLogFileInfo, content);
-            }
-            catch (Exception ex) when (ex is DirectoryNotFoundException || ex is FileNotFoundException)
+        public override void Flush()
+        {
+            Flush(false);
+        }
+
+        private protected virtual bool Flush(bool fromTimer)
+        {
+            if (_logBuffer.Count == 0)
             {
-                // need to handle cases where log files or directories might
-                // have been deleted from underneath us
-                SetLogFile();
-                AppendToFile(_currentLogFileInfo, content);
+                if (fromTimer)
+                {
+                    StopFlushTimer();
+
+                    // We just stopped our timer above; as a sanity check,
+                    // we ensure no traces were written to the buffer in the meantime.
+                    // If new traces showed up, we'll keep going and process them; otherwise,
+                    // we'll just return from here as new traces will re-enable our timer.
+                    if (_logBuffer.Count != 0)
+                    {
+                        StartFlushTimer();
+                    }
+                }
+
+                return true;
             }
 
-            // check to see if we need to roll to a new log file
-            _currentLogFileInfo.Refresh();
-            if (LogFileIsOverSize(_currentLogFileInfo))
+            object flushLock = GetFlushLock();
+            if (!Monitor.TryEnter(flushLock))
             {
-                SetLogFile();
+                return false;
```

Comment:
If we're unable to acquire the lock, we return immediately, without waiting and holding on to the thread.

Reply:
Now that you've scoped the lock to a single file, if you can't enter the monitor here, it means another thread is currently flushing to the file. However isn't returning here without waiting for the lock and writing the logs potentially lossy? E.g. say an "orphaned" host that is spinning down is logging while the new host that is spinning up is also logging. We want all the logs.

Reply:
Yeah, that's possible. One approach would be to make explicit `Flush` calls block and wait here. That would address the issue and still avoid contention in the timer callback scenario, which was where we were seeing issues.

Reply:
Making the change described above would also help remove some of the test specific code I had to add here.